src/memory_manager.rs

```diff
@@ -287,13 +397,96 @@ impl<M: Memory> MemoryManagerInner<M> {
             bucket_size_in_pages: header.bucket_size_in_pages,
             memory_sizes_in_pages: header.memory_sizes_in_pages,
             memory_buckets,
+            freed_buckets: BTreeSet::new(),
+        }
+    }
+
+    fn load_layout_v2(memory: M, header: Header) -> Self {
+        let mut memory_size_in_buckets = vec![];
+        let mut number_of_used_buckets = 0;
+
+        for memory_size_in_pages in header.memory_sizes_in_pages.into_iter() {
+            let size_in_buckets = (memory_size_in_pages + header.bucket_size_in_pages as u64 - 1)
+                / header.bucket_size_in_pages as u64;
+            memory_size_in_buckets.push(size_in_buckets);
+            number_of_used_buckets += size_in_buckets;
+        }
+
+        const BYTE_SIZE_IN_BITS: usize = 8;
+        let buckets_index_size_in_bytes: usize =
+            (number_of_used_buckets as usize * BUCKET_ID_LEN_IN_BITS + (BYTE_SIZE_IN_BITS - 1))
+                / BYTE_SIZE_IN_BITS;
+
+        let mut buckets = vec![0; buckets_index_size_in_bytes];
+        memory.read(bucket_indexes_offset().get(), &mut buckets);
+
+        let buckets_decompressed = bytes_to_bucket_indexes(&buckets);
```

Comment:
One way to make code more easy to read is by scoping. For example, this block can be restructured to be like this:

```suggestion
        // Load the buckets.
        let buckets = {
            const BYTE_SIZE_IN_BITS: usize = 8;
            let buckets_index_size_in_bytes: usize =
                (number_of_used_buckets as usize * BUCKET_ID_LEN_IN_BITS + (BYTE_SIZE_IN_BITS - 1))
                    / BYTE_SIZE_IN_BITS;

            let mut buckets = vec![0; buckets_index_size_in_bytes];
            memory.read(bucket_indexes_offset().get(), &mut buckets);
            bytes_to_bucket_indexes(&buckets)
        };
```

In this case, it's clear that all this logic is trying to load buckets and, as a reader, I don't need to look inside this block unless I want to see how buckets are loaded.